dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorSpirvOutput.java

```diff
@@ -99,38 +102,36 @@ private Expression acceptAssertionValue(SpirvParser.AssertionValueContext ctx) {
         if (base == null) {
             throw new ParsingException("Uninitialized location %s", name);
         }
-        TerminalNode offset = ctx.ModeHeader_PositiveInteger();
-        if (offset == null) {
+        if(ctx.indexValue().isEmpty()) {
             return new Location(name, base, 0);
         }
-        Type type = builder.getVariableType(name);
-        int ind = Integer.parseInt(offset.getText());
-        List<Type> types = unrollType(type);
-        if (ind >= types.size()) {
-            throw new ParsingException("Index out of bounds: %s", offset.getText());
-        }
-        int o = 0;
-        for (int i = 0; i < ind; i++) {
-            o += TYPE_FACTORY.getMemorySizeInBytes(types.get(i));
+        List<Integer> indexes = new ArrayList<>();
+        for (SpirvParser.IndexValueContext index : ctx.indexValue()) {
+            indexes.add(acceptIndexValue(index));
         }
-        return new Location(name + "[" + offset.getText() + "]", base, o);
+        Type type = builder.getVariableType(name);
+        int offset = getOffset(type, indexes);
+        String offsetName = name + "[" + String.join("][", indexes.stream().map(Object::toString).toArray(String[]::new)) + "]";
+        return new Location(offsetName, base, offset);
     }
 
-    private List<Type> unrollType(Type type) {
-        if (type instanceof ArrayType arrayType) {
-            List<Type> result = new ArrayList<>();
-            for (int i = 0; i < arrayType.getNumElements(); i++) {
-                result.addAll(unrollType(arrayType.getElementType()));
-            }
-            return result;
-        } else if (type instanceof AggregateType aggregateType) {
-            List<Type> result = new ArrayList<>();
-            for (Type fieldType : aggregateType.getDirectFields()) {
-                result.addAll(unrollType(fieldType));
+    private int getOffset(Type type, List<Integer> indexes) {
+        int offset = 0;
+        for (int index : indexes) {
```

Comment:
This looks good for a valid input, but I think we should also validate the length and the value of indexes. 

Here is an example test for illegal indexes in a vector (the only correct one would be "%var[0][0]"). I would also add a test for a structure and for using indexes with for a scalar type.

@Test
    public void testIllegalVectorIndex() {
        // TODO: The first case does throw an exception, 
        //  but it would be nice to make it more informative, 
        //  maybe something like "Illegal assertion for variable '%myvar', index out of bounds"
        doTestIllegalVectorIndex("%var[0][0][0]==0",
                "Unexpected index for type bv64");
        doTestIllegalVectorIndex("%var[0]==0", "Some informative error");
        doTestIllegalVectorIndex("%var[1][0]==0", "Some informative error");
        doTestIllegalVectorIndex("%var[0][1]==0", "Some informative error");
    }

    private void doTestIllegalVectorIndex(String ast, String error) {
        String input = """
            ; @Input: %var = {{0}}
            ; @Output: forall (<<ast>>)
                           OpCapability Shader
                    %ext = OpExtInstImport "GLSL.std.450"
                           OpMemoryModel Logical GLSL450
                           OpEntryPoint GLCompute %main "main"
                           OpSource GLSL 450
                   %void = OpTypeVoid
                   %func = OpTypeFunction %void
                 %uint64 = OpTypeInt 64 0
                %v1_type = OpTypeVector %uint64 1
                %v2_type = OpTypeVector %v1_type 1
               %ptr_type = OpTypePointer Uniform %v2_type
                    %var = OpVariable %ptr_type Uniform
                   %main = OpFunction %void None %func
                  %label = OpLabel
                           OpReturn
                           OpFunctionEnd
                 """.replace("<<ast>>", ast);

        try {
            // when
            localParse(input);
            fail("Should throw exception");
        } catch (ParsingException e) {
            // then
            assertEquals(error, e.getMessage());
        }
    }

Reply:
Added, thanks!